docs/security-self-assessment.md

```diff
@@ -0,0 +1,268 @@
+
+OpenFGA has an active community, with 500+ Discord members, 40 unique contributors, and an average of [850 commits per month](https://openfga.devstats.cncf.io/d/2/commits-repository-groups?orgId=1&var-period=m&var-repogroups=All&from=now-1y%2Fy&to=now-1y%2Fy). 
+
+### Actors
+
+**OpenFGA Server**
+
+The OpenFGA server is responsible for storing and querying relationship tuples and authorization models.
+
+Given it's going to be used to know if a user can perform an action on a resource, any potential bug in the OpenFGA server logic can have security implications. 
+
+The server can be configured without authentication, with a shared key, or using the OAuth client credentials flow.
+
+**Database Server**
+
+The database server stores relationship tuples, authorization models and a changelog. OpenFGA currently supports Postgres and MySQL, and it can be extended to support other databases.
+
+Database credentials can be provided as a parameter to the OpenFGA server, via environment variables, or a configuration file.
+
+**CLI/API Clients**
+
+The CLI/API clients are used to make API requests to the OpenFGA server. This includes functionality such as creating and querying relationship tuples, and updating the authorization model.
+
+They authenticate using any of the mechanisms supported by the server (no authentication, shared key, OAuth client credentials).
+
+If the credentials were compromised, the attacker would be able to perform any action granted to those credentials. This would include: 
+
+* Changing the authorization model.
+* Adding tuples to give the attacker access to a resource.
+* Removing tuples to prevent a user from accessing a resource.
+* Querying authorization data, giving the attacker access to privileged information.
+
+To achieve this the attacker would need to be able to connect directly to the OpenFGA service, which should not be accessible from the public internet.
+
+OpenFGA provides versioned authorization models, and applications should point to a specific version. This implies that even if the attacker updates the authorization model, it should not have any effect unless they can also update the application configuration to make it point to the new version.
+
+### Actions
+
+**Invoking the OpenFGA APIs**
+
+Every time a server endpoint is invoked, OpenFGA validates that:
+
+  - The credentials provided in the API call match the ones configured in the server.
+
+  - The payload of the API call:
+    - Matches the Protobuf API definitions.
+    - Validates the parameters have the proper structure, e.g. users need to be written this way 'user:<userid>'
```

Comment:
- Ensures that the input is a semantically valid, e.g. that a tuple is valid according to the authorization model or that the model does not have disallowed cyclical or problematic definitions